### PR TITLE
Clippy

### LINKED ISSUE #42: Additional Lints
@vic1707, you mentioned the potential to add a handful more lints. I assumed just turning on the whole set of pedantic lints was good enough, but sounds like that's not the case.

I did turn on additional lints to remind myself to go add type and module documentation to the `view` module, but I would really prefer to treat warnings as errors so I'm thinking I'll remove that until I've actually added the documentation.

(also as a side note, I did see the nightly requirement is no longer necessary and removed it)

### PR DESCRIPTION
**Pull request by vic1707**
Fixes #42 (?)

Had a bit of fun checking lints and all assuming you'd want to at least see what it looks like ^^
As you said enabling `clippy::pedantic` didn't trigger much (only 2 lints about numbers casts).

Feel free to checkout the branch and see what lints should be enabled (and thus fixed) or if any actual changes bother you so we can revert and bypass the lint.

Todo:
- [ ] : re-organize the bypass lints into categories 
- [ ] : see what I can fix without changing existing behaviours

### CODE REVIEW

File: .cargo/config.toml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+[build]
```
Comment (riley-williams) [inline .cargo/config.toml:1]: Is it better to configure this with rustflags or in the Cargo.toml? I've used the Cargo.toml option to warn on pedantic as well as disabling a couple temporarily. 

I tried enabling the nursery lints in another project and ended up turning it off because a lot ended up just being annoying. I assume a lot of these that have been manually allowed are probably nursery too?

File: tests/match_view.rs [hunk 2]
```diff
@@ -139,17 +141,17 @@ fn test_match_view_two_variants_invalid_layout() {
     let font = CharacterBufferFont;
     let make_view = |state| {
         match_view!(state => {
-            0 => Text::str("zero\n!!!", &font),
+            0i32 => Text::str("zero\n!!!", &font),
```
Comment (riley-williams) [inline tests/match_view.rs:144]: Undecided on this one

File: src/view/hstack.rs [hunk 3]
```diff
@@ -244,7 +249,8 @@ macro_rules! impl_layout_for_hstack {
         impl<$($type: Renderable<C>),+, C> Renderable<C> for HStack<($($type),+)> {
             type Renderables = ($($type::Renderables),+);
 
-            #[allow(unused_assignments)]
+            #[inline]
+            #[expect(unused_assignments)]
```
[the remark this answers is not in the document]
Comment (riley-williams) [reply src/view/hstack.rs:253]: This specific allow should probably also just be moved to the `if` block containing the real issue. Seems like that doesn't work if it's written as #[expect]

File: Cargo.toml [hunk 4]
```diff
@@ -14,7 +14,7 @@ exclude = [".github"]
 maintenance = { status = "experimental" }
 
 [features]
-default = ["embedded-graphics"]
+default = ["embedded-graphics", "crossterm"]
```
[the remark this answers is not in the document]
Comment (vic1707) [reply Cargo.toml:17]: This shouldn't have been committed, that's my-bad. 
My editor doesn't lint or treat the code with the feature disabled.
I didn't look for an actual solution to this yet, to my knowledge a `.vscode/settings.json` and equivalents for other editors does the trick. I'll look into it to avoid committing that by mistake again 😓.

As for the feature themselves I'm not an expert, I'd maybe say no features by default (a compile error can be triggered if no feature are enabled if you want). 
As someone probably wants one of the 2 strategies, rarely both. 
Setting default to `embedded-graphics` means that users will also want to use `default-features = false` in their `Cargo.toml` when wanting only `crossterm`.

File: tests/spacer.rs [hunk 5]
```diff
@@ -1,10 +1,11 @@
+#![cfg(test)]
 use buoyant::font::CharacterBufferFont;
-use buoyant::layout::{Layout, LayoutDirection};
+use buoyant::layout::{Layout as _, LayoutDirection};
```
[the remark this answers is not in the document]
Comment (vic1707) [reply tests/spacer.rs:3]: clippy is fine without this line, but rust-analyzer isn't.
despite this file being in the `tests` folder it's sometimes not recognized as such and thus triggers an error.
I think this issue https://github.com/rust-lang/rust-analyzer/issues/7225 refers to it.
In personal projects I usually put a `lib.rs` in the root of `tests` and list modules there, it seemed to fix the issue (see: https://github.com/vic1707/nnn/blob/main/tests/lib.rs)

File: tests/animation.rs [hunk 6]
```diff
@@ -517,6 +518,7 @@ fn toggle_stack(is_on: bool) -> impl Renderable<char, Renderables: CharacterRend
 }
 
 #[test]
+#[expect(clippy::cognitive_complexity)]
```
[the remark this answers is not in the document]
Comment (vic1707) [reply tests/animation.rs:521]: noted ✅

File: src/view/text/character.rs [hunk 7]
```diff
@@ -1,3 +1,4 @@
+extern crate alloc;
```
[the remark this answers is not in the document]
Comment (vic1707) [reply src/view/text/character.rs:1]: In this case I'd suggest doing an `alloc` feature for the same reason the `std` feature exists

File: tests/match_view.rs (hunk 2 above)
Comment (vic1707) [reply tests/match_view.rs:144]: yeah, it looks weird, a solution would be to inject the `#[allow()]` attribute inside the macro

File: src/view/foreach.rs [hunk 8]
```diff
@@ -152,7 +156,7 @@ where
             let view = (self.build_view)(&item);
             // TODO: If we include an ID here, rows can be animated and transitioned
             let item = renderables.push(view.render_tree(item_layout, aligned_origin, env));
-            assert!(item.is_ok());
+            assert!(item.is_ok(), "");
```
[the remark this answers is not in the document]
Comment (vic1707) [reply src/view/foreach.rs:159]: If you have a suggestion on that specific message 😉.

I'll see how I can remove it. For optimizations I can't answer for sure. I often see `debug_assert!` on heavily tested code to not have to deal with that question.

File: src/view/foreach.rs [hunk 9]
```diff
@@ -113,22 +117,22 @@ where
...
+        let size = layout_n(&subview_stages, *offer, self.spacing, layout_fn);
         ResolvedLayout {
             sublayouts,
             resolved_size: size,
         }
     }
 }
 
-impl<const N: usize, I: IntoIterator + Copy, V: Renderable<C>, C, F> Renderable<C>
+impl<const N: usize, I: IntoIterator + Copy, V: Renderable<C> + Layout, C, F> Renderable<C>
```
[the remark this answers is not in the document]
Comment (vic1707) [reply src/view/foreach.rs:128]: yup, clippy didn't like having bounds in the generic declaration and a `where` clause in the same context, it's one or the other 🙃

File: .cargo/config.toml (hunk 1 above)
Comment (vic1707) [reply .cargo/config.toml:1]: I don't know if there is an accepted/idiomatic answer.
IMO `Cargo.toml` is included in the library publishing process and users read it, I feel like the crate's clippy config isn't of users concern (as in "users don't care", not "should be hidden") so we shouldn't bother them with it 🤔

File: src/render/animate.rs [hunk 10]
```diff
@@ -230,15 +236,13 @@ mod embedded_graphics_impl {
 
             let new_duration;
             let is_partial;
-            let subdomain;
-            if end_time == Duration::from_secs(0) || domain.app_time >= end_time {
-                // animation has already completed or there was zero duration
+            let subdomain = if end_time == Duration::from_secs(0) || domain.app_time >= end_time {
```
[the remark this answers is not in the document]
Comment (vic1707) [reply src/render/animate.rs:239]: fair enough, clippy is opinionated and this changes most definitely comes from `nursery`. I'll disable it globally 👍

File: src/view/modifier/flex_frame.rs [hunk 11]
```diff
@@ -103,12 +103,14 @@ impl<V: Layout> Layout for FlexFrame<V> {
                 self.min_width,
                 self.max_width.map(Into::into),
             )),
-            ProposedDimension::Compact => match self.ideal_width {
-                Some(ideal_width) => ProposedDimension::Exact(
-                    self.min_width.map_or(ideal_width, |w| w.max(ideal_width)),
-                ),
-                None => ProposedDimension::Compact,
-            },
+            ProposedDimension::Compact => {
```
[the remark this answers is not in the document]
Comment (vic1707) [reply src/view/modifier/flex_frame.rs:106]: fair enough, I think those lints are meant to say "hey there's a new way to do that, you certainly want it right?"
I'll disable it globally 👍

File: src/view/hstack.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vic1707) [reply src/view/hstack.rs:253]: > Does this lint want all `#[allow(x)]` to be converted to `#[expect(x)]`?

Yup, came out recently, `#[expect()]` triggers a warning when it becomes useless, it was also the first to let you provide a reason for disabling the lint `#[expect(unused_assignments, reason = "because")]` (`#[allow(x)]` also does since a few versions). I disabled the reason for the time.

> This specific allow should probably also just be moved to the if block containing the real issue. Seems like that doesn't work if it's written as #[expect]

I'll look into it 👍

File: Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (riley-williams) [reply Cargo.toml:17]: Ah no worries, I've accidentally done the same thing. I think it probably makes sense then for me to leave it as-is.

File: tests/padding.rs [hunk 12]
```diff
@@ -1,18 +1,19 @@
-use std::iter::zip;
+#![cfg(test)]
```
[the remark this answers is not in the document]
Comment (riley-williams) [reply tests/padding.rs:1]: I guess I still don't really understand what the issue is after peeking at that issue, maybe because I just have yet to see it in my personal tools/workflow. Is this an actual problem you see now or is it preemptive?

**Comment by riley-williams** [comment]
I think we'll stick to defining these in the Cargo.toml, mainly because this has advantages for workspaces which I imagine this will eventually become.

It makes importing them into the packages trivial, whereas I believe the rust flags must be duplicated for every package.

```toml
# package/Cargo.toml
[lints]
workspace = true
```

I think what I want for the lint selection is actually the inverse. The default lints really go a long way and like you said, they're quite powerful. Rather than dealing with random issues that pop up in unstable nursery and restriction lints, I think I'd rather cherry pick what makes the most sense for this project today.

I went through, and I _think_ this covers the issues you've fixed. I picked out some others that I think make sense, but probably require more involved changes.

```toml
[lints.clippy]
pedantic = { level = "deny", priority = -1 }

# You fixed?
needless_pass_by_ref_mut = "deny"
use_self = "deny"
derive_partial_eq_without_eq = "deny"
too_long_first_doc_paragraph = "deny"
type_repetition_in_bounds = "deny"
option_if_let_else = "deny"
useless_let_if_seq = "deny"

# TODO: Disabled for now, desired fix
cast_possible_wrap = "allow"
cast_possible_truncation = "allow"
exhaustive_structs = "allow"
exhaustive_enums = "allow"
missing_panics_doc = "allow"
field_scoped_visibility_modifiers = "allow"
ref_patterns = "allow"
cast_sign_loss = "allow"
string_slice = "allow"
shadow_unrelated = "allow"
renamed_function_params = "allow"
float_arithmetic = "allow"
indexing_slicing = "allow"
as_conversions = "allow"
absolute_paths = "allow"                    # configure as makes sense
```

**Comment by riley-williams** [comment]
Thanks for all the effort answering my questions in your PR! I (finally) ended up merging the results of this separately in #60 to avoid the conflicts

**Comment by vic1707** [comment]
Hi, I'm terribly sorry you ended up doing it yourself 😞
I just realized how long it's been since I opened my project and buoyant itself 😥
Life is a bit of a mess right now, hopefully I'll come back to your wonderful project asap !

**Comment by riley-williams** [comment]
Life gets busy, no worries. Running `clippy --fix` was the easiest part after figuring out what lints to keep here.
On the bright side, the longer you wait, the better the documentation gets! 
Hope to see you back and that everything settles down soon.
